torch/_dynamo/variables/nn_module.py

```diff
@@ -398,6 +409,66 @@ def generic_call_method_helper(name):
+                from torch._functorch.aot_autograd import create_functional_call
+                from .builtin import BuiltinVariable
+                from .inline_helper import (
+                    decompose_and_inline_function_with_makefx,
+                    reconstruct_node_meta_data,
+                )
+
+                start_node_code = len(tx.output.graph.nodes)
+
+                # For single step graph we want to use make_fx to extract the fx graph
+                # representing the fwd and inline that fx graph.
+                # Dynamo needs to see all inputs and output of each torch function
+                # in case they are saved for the backward.
+                mod = tx.output.get_submodule(self.module_key)
+                fn = mod.forward.__func__
+                # TODO(JackCaoG): considering using source and regular builder instead of
+                # SourcelessBuilder.
+                fn_source = AttrSource(self.source, "__call__")
+                fn_source = AttrSource(fn_source, "__func__")
+
+                # Create the functional version of the module which also lifted
+                # all parameters as inputs.
+                params = {
+                    **dict(mod.named_parameters(remove_duplicate=False)),
+                    **dict(mod.named_buffers(remove_duplicate=False)),
+                }
+                params_flat, params_spec = pytree.tree_flatten(params)
+                params_flat = list(params_flat)
+                params_len = len(params_flat)
+                functional_call = create_functional_call(
+                    mod, params_spec, params_len, check_tuple_output=False
+                )
+
+                # Create VariableTrackers for this module's named_parameters and named_buffers.
+                self_params_as_vt = []
+                for full_name in params_spec.context:
+                    current_obj = self
+                    # handle the multi level named_parameter like `mod.y.z`
+                    for name in full_name.split("."):
+                        current_obj = BuiltinVariable(getattr).call_function(
+                            tx, [current_obj, ConstantVariable.create(name)], {}
+                        )
+                    self_params_as_vt.append(current_obj)
+
+                # Concatenate args and pass it to decompose_and_inline_function_with_makefx
+                complete_tensor_variable_args = self_params_as_vt + args
+                res = decompose_and_inline_function_with_makefx(
+                    tx,
+                    functional_call,
+                    complete_tensor_variable_args,
+                    kwargs,
+                )
+
+                num_nodes_need_update_metadata = (
+                    len(tx.output.graph.nodes) - start_node_code
+                )
+                reconstruct_node_meta_data(self, tx, num_nodes_need_update_metadata)
+                return res
```

Comment:
Another design question. This code path is triggered only iff we have something like this

~~~
class MyMod(torch.nn.Module):
      def forward(self, x):
           return self.submod.forward(x)
~~~

i.e., only when user explicitly calls `forward`. Is this intentional?

Reply:
I think this works when user call module directly, for example in my mini test script I have
```
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch_xla.utils.utils as xu

class MNIST(nn.Module):

  def __init__(self):
    super(MNIST, self).__init__()
    self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
    self.dropout = nn.Dropout(p=0.5, inplace=True)

  def forward(self, x):
    x = F.relu(F.max_pool2d(self.conv1(x), 2))
    x = self.dropout(x)
    return x
  
......


mnist = MNIST()
mnist_compiled = torch.compile(mnist, backend="eager")

for data, target in loader:
  #pred = mnist(data)
  pred_2 = mnist_compiled(data)
```

I think the way it works is that in `NnModuleVaribale.call_function` it will inline the module's fwd in https://github.com/pytorch/pytorch/pull/123683/files#diff-890f369c8fb7fd76d79bac7fe4f8dddf126916580d0a9ee6b64dad221ba3d46eR362-R366

Reply:
Ah, I see .. that makes sense.